### PR TITLE
dahdi-linux: bump + fixes for linux>=5.18 and 6.1 build

### PR DESCRIPTION
**Pull request by john-tho**
Maintainer: 
Compile tested: aarch64 kernel 6.1rc1 WIP
Run tested: no

wait for upstream changes PR review: https://github.com/asterisk/dahdi-linux/pull/14

### CODE REVIEW

**Comment by micmac1** [comment]
Hi there,

is this maybe already addressed in the "next" branch?

https://github.com/asterisk/dahdi-linux/commits/next

**Comment by john-tho** [comment]
> Hi there, is this maybe already addressed in the "next" branch? https://github.com/asterisk/dahdi-linux/commits/next

oops… It does have the pci_ removed fix, but not the `netif_napi_add` fix yet.

switched to git source + next branch + 6.1 fix
not for merge, just for testing/comment for now

Cheers

**Comment by micmac1** [comment]
Hi @john-tho

Can we bump to 3.2.0 instead of going for next branch? And change PKG_RELEASE to "1" (that's where OpenWrt starts counting).

Kind regards,
Seb

**Comment by john-tho** [comment]
switched to the 3.2.0 release + backport 5.18 pci fix + pending 6.1 fix.
Only compile tested 5.15 mt7621. Need to setup my 6.1 branch later to compile test that.

**Comment by micmac1** [comment]
> switched to the 3.2.0 release + backport 5.18 pci fix + pending 6.1
> fix.  Only compile tested 5.15 mt7621. Need to setup my 6.1 branch
> later to compile test that.

Thanks! I have openwrt-sdk-bcm27xx-bcm2710_gcc-12.3.0_musl.Linux-x86_64
SDK here so will test compile against 6.1 tonight.

**Comment by john-tho** [comment]
Added 2 more fixes I ran into on mt7621 6.1 compile:
`dev_addr` const & redefined ABS
and quilt refreshed

**Comment by micmac1** [comment]
Thanks @john-tho 

Compiles fine for me, tried both a kernel 6.1 SDK and a kernel 5.15 SDK.

```
-rw-r--r-- 1 sk sk 3,4K 14. Jun 20:55 bin/targets/bcm27xx/bcm2710/packages/kmod-dahdi-dummy_6.1.33+3.2.0-1_aarch64_cortex-a53.ipk
-rw-r--r-- 1 sk sk 2,6K 14. Jun 20:55 bin/targets/bcm27xx/bcm2710/packages/kmod-dahdi-echocan-oslec_6.1.33+3.2.0-1_aarch64_cortex-a53.ipk
-rw-r--r-- 1 sk sk  12K 14. Jun 20:55 bin/targets/bcm27xx/bcm2710/packages/kmod-dahdi-hfcs_6.1.33+3.2.0-1_aarch64_cortex-a53.ipk
-rw-r--r-- 1 sk sk  63K 14. Jun 20:55 bin/targets/bcm27xx/bcm2710/packages/kmod-dahdi_6.1.33+3.2.0-1_aarch64_cortex-a53.ipk

-rw-r--r-- 1 sk sk 3,2K 14. Jun 21:11 bin/targets/ath79/generic/packages/kmod-dahdi-dummy_5.15.114+3.2.0-1_mips_24kc.ipk
-rw-r--r-- 1 sk sk 2,3K 14. Jun 21:11 bin/targets/ath79/generic/packages/kmod-dahdi-echocan-oslec_5.15.114+3.2.0-1_mips_24kc.ipk
-rw-r--r-- 1 sk sk  12K 14. Jun 21:11 bin/targets/ath79/generic/packages/kmod-dahdi-hfcs_5.15.114+3.2.0-1_mips_24kc.ipk
-rw-r--r-- 1 sk sk  63K 14. Jun 21:11 bin/targets/ath79/generic/packages/kmod-dahdi_5.15.114+3.2.0-1_mips_24kc.ipk
```

I'm not  big on C code, as far as I can tell this looks fine to me. @john-tho are you OK with the current state of your pull request?

@jslachta Hi Jiri, are you OK with me merging this? OpenWrt promoted kernel 6.1 to standard kernel for a few targets, so compile failures are visible with dahdi-linux now. Hence I'd like to merge this soonish. :-)

Thanks all!

Seb

**Comment by jslachta** [comment]
@micmac1, sure!

File: libs/dahdi-linux/patches/206-dahdi-fix-const-dev_addr.patch [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+--- a/drivers/dahdi/datamods/hdlc_raw_eth.c
++++ b/drivers/dahdi/datamods/hdlc_raw_eth.c
+@@ -98,8 +98,8 @@ int hdlc_raw_eth_ioctl(struct net_device
+ 		ether_setup(dev);
+ 		dev->change_mtu = old_ch_mtu;
+ 		dev->tx_queue_len = old_qlen;
+-		memcpy(dev->dev_addr, "\x00\x01", 2);
+-                get_random_bytes(dev->dev_addr + 2, ETH_ALEN - 2);
++		eth_hw_addr_random(dev);
++		dev_addr_mod(dev, 0, "\x00\x01", 2);
```
Comment (john-tho) [inline libs/dahdi-linux/patches/206-dahdi-fix-const-dev_addr.patch:65]: These now are the only changes I am not confident with. They _look like_ they should work, but I might ping @Ansuel for a second opinion.
https://elixir.bootlin.com/linux/latest/C/ident/dev_addr_mod
```c
void dev_addr_mod(struct net_device *dev, unsigned int offset,
		  const void *addr, size_t len);
```

**Comment by pkgadd** [comment]
I've been falling over this while working on bumping x86_64 to kernel v6.1 as well. In regards to kernel fixes, there seems to be quite some activity at https://gitea.osmocom.org/retronetworking/dahdi-linux/ (I've never used dadhi, just compile (tried to-) compile tested it with kernel v6.1).

**Comment by micmac1** [comment]
Hi @john-tho 

I've looked around at other projects and how they solved this. I found a buildroot patch, which buildroot seems to have gotten from a gentoo pull request: [Link](https://github.com/buildroot/buildroot/commit/840b5527466502f02b86c95f9bf95557d35c8571)

Actually replacing your 206 patch directly with the one from Gentoo doesn't work. I get this:

```
/home/sk/tmp/sdk/openwrt-sdk-bcm27xx-bcm2710_gcc-12.3.0_musl.Linux-x86_64/build_dir/target-aarch64_cortex-a53_musl/linux-bcm27xx_bcm2710/dahdi-linux-3.2.0/drivers/dahdi/xpp/xbus-sysfs.c:466:19: error: initialization of 'int (*)(struct device *, struct kobj_uevent_env *)' from incompatible pointer type 'int (*)(const struct device *, struct kobj_uevent_env *)' [-Werror=incompatible-pointer-types]
  466 |         .uevent = astribank_uevent,
      |                   ^~~~~~~~~~~~~~~~
```

But when I just take a single line of that patch it compiles:

```
cat 206-dahdi-fix-const-dev_addr.patch 
--- a/drivers/dahdi/wctc4xxp/base.c
+++ b/drivers/dahdi/wctc4xxp/base.c
@@ -643,7 +643,7 @@ wctc4xxp_net_register(struct wcdte *wc)
 		return -ENOMEM;
 	priv = netdev_priv(netdev);
 	priv->wc = wc;
-	memcpy(netdev->dev_addr, our_mac, sizeof(our_mac));
+	memcpy((void *)netdev->dev_addr, our_mac, sizeof(our_mac));
 
 #	ifdef HAVE_NET_DEVICE_OPS
 	netdev->netdev_ops = &wctc4xxp_netdev_ops;
```

Does this make sense to you?

**Comment by CharlesMengCA** [comment]
Hi john-tho, 

Could you rebase your branch with the master?

Thanks,

**Comment by john-tho** [comment]
> ```
> cat 206-dahdi-fix-const-dev_addr.patch 
> --- a/drivers/dahdi/wctc4xxp/base.c
> +++ b/drivers/dahdi/wctc4xxp/base.c
> @@ -643,7 +643,7 @@ wctc4xxp_net_register(struct wcdte *wc)
>  		return -ENOMEM;
>  	priv = netdev_priv(netdev);
>  	priv->wc = wc;
> -	memcpy(netdev->dev_addr, our_mac, sizeof(our_mac));
> +	memcpy((void *)netdev->dev_addr, our_mac, sizeof(our_mac));
>  
>  #	ifdef HAVE_NET_DEVICE_OPS
>  	netdev->netdev_ops = &wctc4xxp_netdev_ops;
> ```
> Does this make sense to you?

To me, this looks like a dodgy workaround, if the (Linux) intention is that nothing directly modifies `dev_addr`.

Linux 5.17 made `dev_addr` const [adeef3e32146 ("net: constify netdev->dev_addr")](https://github.com/torvalds/linux/commit/adeef3e32146) As part of that series, there are runtime checks against this:  [d07b26f5bbea ("dev_addr: add a modification check")](https://github.com/torvalds/linux/commit/d07b26f5bbea9ade34dfd6abea7b3ca056c03cd1) From https://lore.kernel.org/netdev/20211119142155.3779933-6-kuba@kernel.org/:
>netdev->dev_addr should only be modified via helpers,
but someone may be casting off the const. Add a runtime
check to catch abuses.



---

> Could you rebase your branch with the master?

I cannot see any conflicts, could you please let me know why you want the rebase?

Cheers,

**Comment by micmac1** [comment]
Thanks john-tho!

**Comment by john-tho** [comment]
For a quick update: There has finally been some activity on the dahdi repo. Most of the 6.1 build issues have been addressed in the master branch, and I opened an issue to detail the `dev_addr` direct writes: https://github.com/asterisk/dahdi-linux/issues/31

**Comment by micmac1** [comment]
Closing this, merged bump to head in another PR. @john-tho thanks for your work!

**Pull request closed without merging**